tools/shared/src/main/scala/org/scalajs/core/tools/linker/analyzer/Analyzer.scala

```diff
@@ -165,6 +173,7 @@ private final class Analyzer(semantics: Semantics,
   private class ClassInfo(data: Infos.ClassInfo) extends Analysis.ClassInfo {
     private[this] var _linking = false
```

Comment:
:P No, not really. The problem is more that `_linkingFrom == null` can mean two things then:

- I'm not linking
- I'm a linking root

Not being able to distinguish these cases is problematic. Not because we can't detect cycles, but because we can't reconstruct the cycles.

I think it would be cleaner if you made `_linkingFrom` an `Option` and then essentially have proper tri-state (with the `null`).